### PR TITLE
Bug 1318715 reflect site bookmarked state context menu

### PR DESCRIPTION
**Pull request by fluffyemily**
If a `Site` has a bookmarked state, use that to toggle bookmark on/off, otherwise fetch bookmarked state from DB, add to `Site` and use that instead.

One thing I don't like - opening up `bookmarked` in `Site` for editing but I couldn't think of another way....

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
ButtonToast.swift | 18 | Colons should be next to the identifier when specifying a type. 
#### Errors

File | Line | Reason |
| --- | ----- | ----- |
ButtonToast.swift | 23 | Opening braces should be preceded by a single space and on the same line as the declaration. 
ScreenGraph.swift | 53 | Opening braces should be preceded by a single space and on the same line as the declaration. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: Client/Frontend/Home/ActivityStreamPanel.swift [hunk 1]
```diff
@@ -404,6 +404,22 @@ extension ActivityStreamPanel {
...
+                        }
+
+                        site.setBookmarked(isBookmarked)
+                        dispatch_async(dispatch_get_main_queue()) {
+                            self.presentContextMenu(site, eventInfo: eventSource, siteImage: siteImage, siteBGColor: siteBGColor)
+                        }
+                }
+            }
+        }
         presentContextMenu(site, eventInfo: eventSource, siteImage: siteImage, siteBGColor: siteBGColor)
```
Comment (farhanpatel) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: There is no else block here. So I think this might present the context menu twice.

**Comment by farhanpatel** [comment]
Haha. There is some weirdness going on here. Bryan already had a [patch](https://github.com/mozilla-mobile/firefox-ios/pull/2130) for this. But then he reverted it a commit later in [this](https://github.com/mozilla-mobile/firefox-ios/pull/2143) PR? I don't think by purpose either, at least it doesn't mention anything in the PR. wtf is going on lol. This is hilarious  @bkmunar

**Comment by farhanpatel** [comment]
Haha. So according to IRC. Bryan doesn't know what happened 😄 

I kind of liked the way @bkmunar did it though. Getting the bookmark status from the SQL query instead of each time the item is pressed. Although his patch is missing support for TopSites which yours does have. 

sidenote: Any thoughts on how we can abstract this logic to a different class/file. Being able to open the context menu with these items will have to be done from other panels (bookmarks/history) as well. we have a PR for this https://github.com/mozilla-mobile/firefox-ios/pull/2268 But there's a lot of duplicated code. I remember you worked on the Menu. Is there something we can do similar here?

**Comment by farhanpatel** [comment]
Also if you have any suggestions on how to refactor this code (and have time to share your thoughts between the swift 3 stuff) let me know. I know it aint too old but I'm already itching to refactor it haha.

File: Client/Frontend/Home/ActivityStreamPanel.swift (hunk 1 above)
Comment (fluffyemily) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: The `return` before the fetch bookmark status call prevents the context menu being called twice, but I'll move it to make that clearer.

**Comment by fluffyemily** [comment]
@farhanpatel yeah, I liked the addition of bookmark fetching for highlights, but this still doesn't cover the case of bookmarked Default Suggested Sites which are hardcoded by region and don't check for bookmark status anywhere. I was trying to think of a way of doing that as we created the Default Suggested Sites, but couldn't reconcile the async nature of the bookmark status fetch with the fact that we might be accessing the property before the check has completed.

File: Client/Frontend/Home/ActivityStreamPanel.swift [hunk 2]
```diff
@@ -404,6 +404,22 @@ extension ActivityStreamPanel {
...
+            return profile.bookmarks.modelFactory >>== {
+                $0.isBookmarked(site.url)
+                    .uponQueue(dispatch_get_main_queue()) {
+                        guard let isBookmarked = $0.successValue else {
+                            log.error("Error getting bookmark status: \($0.failureValue).")
+                            return
+                        }
+
+                        site.setBookmarked(isBookmarked)
+                        dispatch_async(dispatch_get_main_queue()) {
```
Comment (fluffyemily) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: yeah, I missed that. Good spot. Thanks :)

**Comment by fluffyemily** [comment]
So, it appears that the changes for adding `is_bookmarked` to the Highlights fetching SQL is still present, it was just the checks for bookmark status in `ActivityPanel` that got lost

File: Storage/Site.swift [hunk 3]
```diff
@@ -74,7 +74,7 @@ public class Site: Identifiable {
      // Sites may have multiple favicons. We'll return the largest.
     public var icon: Favicon?
     public var latestVisit: Visit?
-    public let bookmarked: Bool?
+    public private(set) var bookmarked: Bool?
 
     public convenience init(url: String, title: String) {
         self.init(url: url, title: title, bookmarked: false)
```
Comment (farhanpatel) [inline Storage/Site.swift]: when you create a Site. It always has `bookmarked` set to false. But unless `bookmared` is set to nil the ASPanel wont check the bookmark status?

Comment (fluffyemily) [inline Storage/Site.swift]: `basicHistoryColumnFactory` which is used when we fetch Top Sites from the DB, and also used when we fetch anything from history explicitly uses the designated initialiser and a bookmarked value from the DB. The only usage of this convenience initialiser is not in anything used by the `ActivityStreamPanel`. The `DefaultSuggestedSites` which is the case that we are trying to catch with the nil check passes `nil` into the designated initializer: SuggestedSites.swift line 22
```
super.init(url: data.url, title: data.title, bookmarked: nil)
```

**Comment by farhanpatel** [comment]
Ah alright. makes sense.

**Pull request merged**